### ISSUE TITLE
train_ppo.py response_text 截取错误

### ISSUE THREAD

**Issue by SSFelixzyk** [jingyaogong/minimind#640]
在读 train_ppo.py 代码的时候发现 response_text 的截取方式有问题，会截取 prompt。

        prompts = batch["prompt"]  # list[str], length B
        enc = tokenizer(prompts, return_tensors="pt", padding=True, truncation=True, 
                max_length=args.max_seq_len).to(args.device)  # input_ids: [B, P], attention_mask: [B, P]
        prompt_lengths = enc.attention_mask.sum(dim=1) 

        with torch.no_grad():
            model_for_gen = ....
            gen_out = model_for_gen.generate(....)

        responses_text = [tokenizer.decode(gen_out[i, prompt_lengths[i]:], skip_special_tokens=True) for i in range(len(prompts))

这里 tokenizer 使用的是 left padding

        tokenizer.padding_side = 'left'  # PPO需要左侧padding

而 prompt_lengths 是用 有效 token （非padding）的长度， 导致 responses_text 会切到部分 prompt
例如：

        prompt = [pad, pad, p1, p2 p3]
        prompt_lengths = 3
        gen_out = [pad, pad, p1, p2, p3, r1, r2, r3, .....]
        responses_text = gen_out[ 3 : ] = [p2. p3, r1, r2, ....]

**Comment by jingyaogong** [comment]
https://github.com/jingyaogong/minimind/blob/3a5aba82dbecb8b31349340f0e7cf8df94b9aa38/trainer/train_ppo.py#L127

假设 batch 中最长 prompt 有 5 个 token，left padding 后：
```
prompt = [pad, pad, p1, p2, p3]   # shape[1] = 5
gen_out = [pad, pad, p1, p2, p3, r1, r2, r3, ...]
```
用 prompt_length = shape[1] = 5 截取：
```
responses_text = gen_out[5:] = [r1, r2, r3, ...]  # 正确
```
用有效 token 数量 3 来截取：
```
responses_text = gen_out[3:] = [p2, p3, r1, r2, ...]  # 这确实有问题
```

所以现在的逻辑是正确的，你可能误读了变量名和实现方式（或者你的代码没有更新？我记得这是比较久之前存在的问题了）

**Comment by SSFelixzyk** [comment]
确实，应该是代码没有更新，用 shape[1] 是对的
